File: 161_one_edit_distance.java
/*
Given two strings s and t, return true if they are both one edit distance apart, otherwise return false.

A string s is said to be one distance apart from a string t if you can:

Insert exactly one character into s to get t.
Delete exactly one character from s to get t.
Replace exactly one character of s with a different character to get t.
 

Example 1:

Input: s = "ab", t = "acb"
Output: true
Explanation: We can insert 'c' into s to get t.
Example 2:

Input: s = "", t = ""
Output: false
Explanation: We cannot get t from s by only one step.
Example 3:

Input: s = "a", t = ""
Output: true
Example 4:

Input: s = "", t = "A"
Output: true
 

Constraints:

0 <= s.length <= 104
0 <= t.length <= 104
s and t consist of lower-case letters, upper-case letters and/or digits.
*/

class Solution {
    public boolean isOneEditDistance(String s, String t) {
        int ns = s.length();
        int nt = t.length();
        
        if (ns > nt)
            return isOneEditDistance(t, s);
        
        if (nt - ns > 1)
            return false;
        
        for (int i = 0; i < ns; i++) 
            if (s.charAt(i) != t.charAt(i))
                if (ns == nt)
                    return s.substring(i+1).equals(t.substring(i+1));
                else
                    return s.substring(i).equals(t.substring(i+1));
        
        return (ns + 1 == nt);
    }
}

File: 161_one_edit_distance.py
'''
Given two strings s and t, determine if they are both one edit distance apart.

Note: 

There are 3 possiblities to satisify one edit distance apart:

Insert a character into s to get t
Delete a character from s to get t
Replace a character of s to get t
Example 1:

Input: s = "ab", t = "acb"
Output: true
Explanation: We can insert 'c' into s to get t.
Example 2:

Input: s = "cab", t = "ad"
Output: false
Explanation: We cannot get t from s by only one step.
Example 3:

Input: s = "1203", t = "1213"
Output: true
Explanation: We can replace '0' with '1' to get t.
'''

class Solution:
    def isOneEditDistance(self, s: str, t: str) -> bool:
        if s == t:
            return False
        
        ns, nt = len(s), len(t) 
        if ns > nt:
            s, t = t, s
            ns, nt = nt, ns
        
        if nt - ns > 1:
            return False
        
        si, ti = 0, 0
        diff_found = False
        while si < ns and  ti < nt:
            if s[si] == t[ti]:
                si += 1
                ti += 1
            else:
                if diff_found:
                    return False
                diff_found = True
                if ns < nt:
                    ti += 1
                else:
                    si += 1
                    ti += 1
        return True

